server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1510,99 @@ public void testSnapshotFromStreamInput() throws IOException {
+        }
+        assertThat(expectedSeqNo, hasSize(0));
+    }
+
+    public void testRandomExceptionsOnTrimOperations( ) throws Exception {
+        Path tempDir = createTempDir();
+        final FailSwitch fail = new FailSwitch();
+        fail.failNever();
+        TranslogConfig config = getTranslogConfig(tempDir);
+        List<FileChannel> fileChannels = new ArrayList<>();
+        final Translog failableTLog =
+            getFailableTranslog(fail, config, randomBoolean(), false, null, createTranslogDeletionPolicy(), fileChannels);
+
+        expectThrows(IOException.class,
+            () -> {
+                int translogOperations = 0;
+                int maxAttempts = randomIntBetween(5, 10);
+                for(int attempt = 0; attempt < maxAttempts; attempt++) {
+                    int maxTrimmedSeqNo;
+                    try {
+                        fail.failNever();
+                        int extraTranslogOperations = randomIntBetween(10, 100);
+
+                        List<Translog.Index> operations = new ArrayList<>();
+                        for (int op = translogOperations; op < translogOperations + extraTranslogOperations; op++) {
+                            String ascii = randomAlphaOfLengthBetween(1, 50);
+                            Translog.Index operation = new Translog.Index("test", "" + op, op,
+                                primaryTerm.get(), ascii.getBytes("UTF-8"));
+                            operations.add(operation);
+                        }
+                        Randomness.shuffle(operations);
+
+                        translogOperations += extraTranslogOperations;
+                        maxTrimmedSeqNo = translogOperations - randomIntBetween(4, 8);
+
+                        for (Translog.Index operation : operations) {
+                            failableTLog.add(operation);
+                        }
+
+                        // at least one roll + inc of primary term has to be there - otherwise trim would not take place at all
+                        if (attempt == 0 || randomBoolean()) {
+                            primaryTerm.incrementAndGet();
+                            failableTLog.rollGeneration();
+                        }
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+
+                    fail.failRate(30);
+                    failableTLog.trimOperations(primaryTerm.get(), maxTrimmedSeqNo);
+                }
+            });
+
+        assertThat(fileChannels, is(not(empty())));
+        assertThat("all file channels have to be closed",
+            fileChannels.stream().filter(f -> f.isOpen()).findFirst().isPresent(), is(false));
+
+        final AlreadyClosedException alreadyClosedException = expectThrows(AlreadyClosedException.class, () -> failableTLog.newSnapshot());
+        assertThat(alreadyClosedException.getMessage(),
```

Comment:
can you check that you can reopen the translog and that you can at least read all operations that weren't trimmed? this should not result in a translog corruption.